Widget Visibility: add controls to widget block editor

mreishus opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

Fixes https://github.com/Automattic/jetpack/issues/20057

#### Changes proposed in this Pull Request:
* Re-implement widget visibility for block based widgets
  * Widget visibility is an existing jetpack feature for old-style widget (non-block based)
  * While creating and editing block based widgets, you may set up visibility rules inside the "Advanced" section of the Gutenberg sidebar
  * While visiting the front-end, widgets will hide and show depending on the rules that have been added
  * Note: While old-style widgets only had visibility rules for each widget, this PR allows visibility rules to be set up for each block within each widget, allowing for complex setups

![2021-08-25_14-21](https://user-images.githubusercontent.com/937354/130851964-fc30301b-2d64-4913-adf7-2bb64572ac10.png)
![2021-08-23_09-37](https://user-images.githubusercontent.com/937354/130466373-88d160ef-2503-4083-abf1-34c018881a1f.png)
![2021-08-23_09-37_1](https://user-images.githubusercontent.com/937354/130466388-969de40e-7776-4021-8167-0622ff8bbff1.png)


#### Issues

* **(2) Performance monitoring.** mtias asks "Are we monitoring any performance changes this might introduce?"
  * We are not, and I'm currently not sure how to handle this. 
  * Is there prior art for doing this?


#### Issues (Fixed)

 * ~~`widget_conditions_data` has some nested arrays as options, that render fancy nested selects.~~ Fixed, however it isn't perfect.
   * I was able to make the header options unselectable by disabling them, but gutenberg does not have any components that use `<optgroup` elements. (which is what the legacy visbility section uses)
 * ~~* Major option "Taxonomy" should be hidden if there's not a non-default taxonomy~~
 * ~~user and role major options should be hidden on wpcom~~. Fixed.
 * ~~Does not actually control visibility on the front end.~~ Fixed.
 * ~~Javascript build~~
   * ~~I'd like to write in JSX, but I haven't found a way. So I'm using element.createElement.~~
     * ~~Well, I did see instant search, which uses its own preact+webpack+babel setup, but it looked so complicated that I would probably be spending days just on environment setup. [babel](https://github.com/Automattic/jetpack/blob/2d74cbdeb98d117af8422b569da6ef056627376e/projects/plugins/jetpack/modules/search/instant-search/babel.config.js#L1), [package.json and more](https://github.com/Automattic/jetpack/blob/2d74cbdeb98d117af8422b569da6ef056627376e/projects/plugins/jetpack/package.json#L30-L33)~~
   * ~~I am using ES6 features, but I think my code isn't being run by babel, so I might have to convert it to ES5 (this could be tricky with the immutable updates).~~
   * Implemented Jeremy's build process. Thanks!
 * ~~Technically, legacy widgets are also blocks. This means you can currently specify their visibility on two levels, the legacy level and the new block based one, which would lead to confusing results.~~ Fixed: These are no longer allowed to have controls.
 *  ~~Accessibility issues on the delete button~~
 *  ~~On this branch, the /wp-admin/widgets.php page loads with the footer collapsed. There is an uncaught `Exception { name: "NS_ERROR_FAILURE", message: "", result: 2147500037,` shown in the browser console that's related to this behavior.~~
    * Fixed here: e00a1f5cc. It was happening when we were running `setAttributes` on the `core/legacy-widget` and `core/widget-area` blocks, which don't need visibility settings.
 * ~~I cannot add attributes to these types of blocks: 'core/archives', 'core/latest-comments', 'core/latest-posts'. Most blocks let me add a custom attribute, but with these, they start crashing when I add an attribute to them. Why? And are there others that will crash that I haven't found yet?~~
   * fix is in eb6564329d1baa5dc5185e5c7bbf552ec6aae588 
* ~~**(1) Customizer Bug.** F5 -> Visit customizer -> Widgets -> Select a block widget -> Three dots menu -> Show more settings -> Advanced -> Add new rule.~~
  * Expect to see: A new rule added and you are still looking at the list of rules.
  * Actually see: A new rule added and you are back to looking at the block widget. You must: Three dots menu -> Show more settings -> Advanced -> Add new rule.
  * **This bug is non-unique to the code in this branch.** It's an existing bug. For example, try:   F5 -> Visit customizer -> Widgets -> Select a block widget -> Three dots menu -> Show more settings -> Set the "Font Size" select to Large.  Now you're brought back to the block widget in the same manner.
  * **This bug is caused by code in core WP.org.** Commenting out [these 8 lines](https://github.com/WordPress/wordpress-develop/blame/d4abd5ca4a500f7b704c59d7a03ac1c583fde6a3/src/js/_enqueues/wp/customize/controls.js#L7199-L7206) stops it from happening. They were added in [this commit](https://github.com/WordPress/wordpress-develop/commit/d2540cfa069c23adc2f214bb40a0bb51eee445d3). [Comment in trac thread](https://core.trac.wordpress.org/ticket/37661#comment:216)
    * **Even though code in WP.org is causing it, it does not happen on a standalone .org installation.** This is one of those tricky issues that is happening in an interplay between wp.com and wp.org code.
  * I don't think this bug should be solved in this PR, but elsewhere. Should it be a blocker for this feature?
  * Reported here: https://github.com/Automattic/wp-calypso/issues/55875
  * And on .org trac: https://core.trac.wordpress.org/ticket/54063
  * Created Gutenberg PR which I believe fixes this: https://github.com/WordPress/gutenberg/pull/34543
  * PR will be released in Gutenberg 11.5, scheduled for release on Sept. 15

#### Random Musings

 * If you can set visibility rules on any block, it feels like there are so many different places to set them, it could be a potential foot-gun.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

#### Testing instructions:
* `jetpack docker up`
* Make sure WP is installed and you can visit /wp-admin. You may need to use `jetpack docker install`, and possibly edit one of the env files such that your password isn't easily guessable if you decide to use a tunneling solution in the future.
* In another terminal, `jetpack watch plugins/jetpack`
  * You might also need to `jetpack build plugins/jetpack` first.
* Make sure this setting is turned on in jetpack settings.
![2021-08-17_17-46](https://user-images.githubusercontent.com/937354/129810581-2a3911d3-7670-4843-a08d-5707813ef5da.png)
* Visit http://localhost/wp-admin/widgets.php
* Add a block based widget here
* In any block, go to the advanced section of the gutenberg sidebar

![2021-08-17_17-39](https://user-images.githubusercontent.com/937354/129809985-92438675-d682-4463-9889-b163cef6ad54.png)
^ Here it is with no design (although, it might be challenging to fit in sidebar)

Compare to the legacy visibility which we are re-implementing:
![2021-08-17_17-40](https://user-images.githubusercontent.com/937354/129810046-2a7fa9f4-e81a-4442-8a5b-8379d6d5ea4f.png)
^ Legacy visibility, not in this PR

#### Testing on DotCom
* Sandbox a simple site
* Apply D65619-code

mreishus commented:
@Automattic/jetpack-design 
I'm definitely going to need some help with the design. What I'm doing in this PR is rebuilding the 'widget visibility' settings for block based widgets. This will be placed in the advanced section of the gutenberg sidebar. I believe this means I need to be using gutenberg components, and adhering to their style guidelines as closely as possible.

Here's the current implementation of widget visibility for legacy widgets:
![2021-08-17_17-40](https://user-images.githubusercontent.com/937354/129810842-4571c3b4-c9fd-4b3d-8130-e3d2e94844ed.png)

It's a rule based setting, where you can add and remove individual rules. Each rule has a major and minor portion. (Example: Major = [page] Minor = [Home page]. Complete rule: "When the current page is the homepage...").  The rules can be linked together with ANDs or ORs. And finally, there's a "Hide" or "Show" action to place.

Here's the result of me haphazardly throwing elements together without a care for the design:
![2021-08-17_17-39](https://user-images.githubusercontent.com/937354/129810995-46ba8717-4927-49d1-9c46-8e6f382f4172.png)

Needs a lot of work, and there isn't a lot of space available in the sidebar for complicated controls like these. Any guidance is appreciated, but no rush as I'm still focusing on getting the functionality working.

olaolusoga commented:
how does core handle this?

mreishus commented:
I added code that recursively filters blocks based on their visibility conditions. For example, you could have a widget that's two columns, then set the right column to only display on certain pages. While this code works fine, it creates blocks that fail to render and crash php sometimes.

```json
[{"blockName":"core\/columns","attrs":{"conditions":{"action":"show","rules":[],"match_all":"0"}},"innerBlocks":[{"blockName":"core\/column","attrs":{"conditions":{"action":"show","rules":[],"match_all":"0"}},"innerBlocks":[{"blockName":"core\/paragraph","attrs":{"conditions":{"action":"show","rules":[{"major":"page","minor":"front"}],"match_all":"0"}},"innerBlocks":[],"innerHTML":"\n<p>Home Page Only<\/p>\n","innerContent":["\n<p>Home Page Only<\/p>\n"]}],"innerHTML":"\n<div class=\"wp-block-column\"><\/div>\n","innerContent":["\n<div class=\"wp-block-column\">",null,"<\/div>\n"]},{"blockName":"core\/column","attrs":{"conditions":{"action":"show","rules":[],"match_all":"0"}},"innerBlocks":[],"innerHTML":"\n<div class=\"wp-block-column\"><\/div>\n","innerContent":["\n<div class=\"wp-block-column\">",null,"<\/div>\n"]}],"innerHTML":"\n<div class=\"wp-block-columns\">\n\n<\/div>\n","innerContent":["\n<div class=\"wp-block-columns\">",null,"\n\n",null,"<\/div>\n"]}]
```
^ That's an example $blocks that fails to render. I believe it's because I've wiped out the innerBlocks the column, but the `null` in the innerContent might mean "Render the next child here", but it no longer exists. So I need a better way to control block visibility rather than just deleting the block.
![2021-08-18_18-07](https://user-images.githubusercontent.com/937354/129983437-7d2236a2-f077-46aa-a986-1ee182147e43.png)

simison commented:
Some design examples from other plugins at pbAok1-2dI-p2

cpapazoglou commented:
@dsas could you take a look please cause I will be AFK?

mreishus commented:
I was able to get the recursive block filtering working on the front-end. Here's a quick demo (turn on sound for commentary):


https://user-images.githubusercontent.com/937354/130100848-35bb5473-6996-4378-a9d5-be575e95f509.mp4

mreishus commented:
> how does core handle this?

I think core doesn't have a widget visibility feature, and it's always been a jetpack feature. However I'm not certain.

dsas commented:
There's an issue when using the customizer to edit the widget on a jetpack site. After setting up a widget with at least one visibility rule and toggling the match all / match one, it goes straight back to the widget list, rather than staying on the edit widget screen. The toggle does take effect though. Here's a short video: https://d.pr/v/DDlJwE

mreishus commented:
I'm able to replicate this. For me, it's only bringing me back the first time. If I go to advanced and then toggle the match all, it brings me back to the widget list. But if I go back into the same advanced screen, I'm free to toggle it more without being sent back.

https://user-images.githubusercontent.com/937354/130149922-d762b531-f39c-4303-b35d-431a68b08dc3.mp4
^ Toggling "match_all" on jetpack + this branch | First edit sends me back

I've also noticed it happens when I do other settings, like colors:

https://user-images.githubusercontent.com/937354/130149989-ee169b10-62bd-4546-8ec9-f55ce514bd8a.mp4
^ Changing color on jetpack + this branch | First edit sends me back

I also tried it on a production simple site.  I found that any edit I made in the advanced section sent me back Additionally, it didn't have the "Only first edit" restriction like I was seeing on jetpack.

https://user-images.githubusercontent.com/937354/130150071-fcd80fe6-f7c5-49d4-b7a7-0a8bd4ad2dde.mp4
^ Changing settings on production simple | All edits send me back

Something it definitely going on. I don't think it's caused by any of this code, but I agree it would make working in the rules area very difficult if it continues to work this way, so perhaps we can identify what it is and how to disable it.

dsas commented on projects/plugins/jetpack/modules/widget-visibility/widget-conditions/widget-conditions.js:
```diff
@@ -311,3 +311,328 @@ jQuery( function ( $ ) {
+	return createElement( 'div', null, 'or' );
+};
+
+const visibilityAdvancedControls = wp.compose.createHigherOrderComponent( BlockEdit => {
```
On the classic widgets screen (can be enabled by installing Classic Widgets plugin from core and then going to /wp-admin/widgets.php wp.compose won't be available so this will result in an uncaught TypeError

mreishus commented:
@jeherve Can you help with some build environment questions? In `widget-conditions.js`, can I use [JSX](https://github.com/Automattic/jetpack/blob/ad5251463b02e35bd51a85209aa11a7a0579ad9a/projects/plugins/jetpack/modules/widget-visibility/widget-conditions/widget-conditions.js#L405), [the spread operator](https://github.com/Automattic/jetpack/blob/ad5251463b02e35bd51a85209aa11a7a0579ad9a/projects/plugins/jetpack/modules/widget-visibility/widget-conditions/widget-conditions.js#L465), [let/const](https://github.com/Automattic/jetpack/blob/ad5251463b02e35bd51a85209aa11a7a0579ad9a/projects/plugins/jetpack/modules/widget-visibility/widget-conditions/widget-conditions.js#L462),  and [arrow functions](https://github.com/Automattic/jetpack/blob/ad5251463b02e35bd51a85209aa11a7a0579ad9a/projects/plugins/jetpack/modules/widget-visibility/widget-conditions/widget-conditions.js#L428)? All of these worked for me except JSX, which makes me think that this code isn't being transpiled, and I shouldn't be using these things. I'm guessing I need to remove all ES6 features?

dsas commented:
Works identically on a simple site too 👍 

> I think core doesn't have a widget visibility feature, and it's always been a jetpack feature. However I'm not certain. 

This is correct.

jeherve commented:
> this code isn't being transpiled, and I shouldn't be using these things.

Correct, you're a bit limited with all that at the moment. We only minify the file.

That doesn't mean it has to stay that way though :)

This may be a good opportunity to start using modern code in there, and transpile it. Maybe that could replace the existing Widget visibility script, or be a new entry point that would only handle block controls?

Maybe something like 7292dcebea8fd4c652c8dc4ddde59d66c481eeda ?

mreishus commented:
Thanks, I will try out some things with the new system and possibly bring it over.

mreishus commented:
Wanted to document one strange error happening, unrelated to the build stuff.

On the master branch, the widgets page loads with the footer expanded:
![2021-08-20_10-05](https://user-images.githubusercontent.com/937354/130254141-eccf7d26-0569-4020-a3b6-973ccfc5fc2c.png)

On this branch, the page loads with the footer collapsed:
![2021-08-20_10-06](https://user-images.githubusercontent.com/937354/130254179-6df1c8e0-55a8-47fa-ae99-3151a1fa1588.png)

Additionally, there is a strange error message that does not happen on the master branch:
![2021-08-20_10-06_1](https://user-images.githubusercontent.com/937354/130254260-f48ad2bc-44bb-41cd-bc78-6cf128e2614d.png)

Everything in the "stack trace" points to line 14, as well as the browser's origination point for the exception:
![2021-08-20_10-07](https://user-images.githubusercontent.com/937354/130254281-56625c82-fb09-444c-ac5f-fd0ea56274b3.png)

Unminified, this seems to be some emoji related code(?)
```js
!(function (e, a, t) {
    var n,
        r,
        o,
        i = a.createElement("canvas"),
        p = i.getContext && i.getContext("2d");
    function s(e, t) {
        var a = String.fromCharCode;
        p.clearRect(0, 0, i.width, i.height), p.fillText(a.apply(this, e), 0, 0);
        e = i.toDataURL();
        return p.clearRect(0, 0, i.width, i.height), p.fillText(a.apply(this, t), 0, 0), e === i.toDataURL();
    }
    function c(e) {
        var t = a.createElement("script");
        (t.src = e), (t.defer = t.type = "text/javascript"), a.getElementsByTagName("head")[0].appendChild(t);
    }
    for (o = Array("flag", "emoji"), t.supports = { everything: !0, everythingExceptFlag: !0 }, r = 0; r < o.length; r++)
        (t.supports[o[r]] = (function (e) {
            if (!p || !p.fillText) return !1;
            switch (((p.textBaseline = "top"), (p.font = "600 32px Arial"), e)) {
                case "flag":
                    return s([127987, 65039, 8205, 9895, 65039], [127987, 65039, 8203, 9895, 65039])
                        ? !1
                        : !s([55356, 56826, 55356, 56819], [55356, 56826, 8203, 55356, 56819]) &&
                              !s(
                                  [55356, 57332, 56128, 56423, 56128, 56418, 56128, 56421, 56128, 56430, 56128, 56423, 56128, 56447],
                                  [55356, 57332, 8203, 56128, 56423, 8203, 56128, 56418, 8203, 56128, 56421, 8203, 56128, 56430, 8203, 56128, 56423, 8203, 56128, 56447]
                              );
                case "emoji":
                    return !s([10084, 65039, 8205, 55357, 56613], [10084, 65039, 8203, 55357, 56613]);
            }
            return !1;
        })(o[r])),
            (t.supports.everything = t.supports.everything && t.supports[o[r]]),
            "flag" !== o[r] && (t.supports.everythingExceptFlag = t.supports.everythingExceptFlag && t.supports[o[r]]);
    (t.supports.everythingExceptFlag = t.supports.everythingExceptFlag && !t.supports.flag),
        (t.DOMReady = !1),
        (t.readyCallback = function () {
            t.DOMReady = !0;
        }),
        t.supports.everything ||
            ((n = function () {
                t.readyCallback();
            }),
            a.addEventListener
                ? (a.addEventListener("DOMContentLoaded", n, !1), e.addEventListener("load", n, !1))
                : (e.attachEvent("onload", n),
                  a.attachEvent("onreadystatechange", function () {
                      "complete" === a.readyState && t.readyCallback();
                  })),
            (n = t.source || {}).concatemoji ? c(n.concatemoji) : n.wpemoji && n.twemoji && (c(n.twemoji), c(n.wpemoji)));
})(window, document, window._wpemojiSettings);
```
Not sure what is going on here.

mreishus commented:
I was able to get the build system going and added some basic CSS improvements. It's still not prod ready, but it's better.

![2021-08-20_17-01](https://user-images.githubusercontent.com/937354/130297657-523cf1df-b0c8-4571-b5f4-a97141a63411.png)
![2021-08-20_17-02](https://user-images.githubusercontent.com/937354/130297661-2854f75d-739f-4043-9b26-3587f9b48f13.png)

dsas commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,284 @@
+			<div className="widget-vis__rule-col-1">
+				<div className="widget-vis__flex">
+					<div className="widget-vis__delete-rule">
+						<Button onClick={ onDelete }>
+							<Icon icon={ close } size={ 14 } />
+						</Button>
```
I don't think this will be accessible, a screen reader won't be able to tell what this button does. Adding a label prop will address that.

sixhours commented:
I made some tweaks to the UI, copy, and padding/spacing adjustments to hopefully make this more readable:

No rules yet:
<img width="277" alt="Screen Shot 2021-08-23 at 12 58 57 PM" src="https://user-images.githubusercontent.com/2124984/130487361-66d4f597-ec6c-461a-937c-c22410c8d0e9.png">

A few rules:
<img width="281" alt="Screen Shot 2021-08-23 at 1 02 12 PM" src="https://user-images.githubusercontent.com/2124984/130487586-2e4ce356-e9be-41b8-9158-a57123d5ec1e.png">

mreishus commented:
I think we're pretty much ready except for these bugs:

 * I cannot add attributes to these types of blocks: 'core/archives', 'core/latest-comments', 'core/latest-posts'. Most blocks let me add a custom attribute, but with these, they start crashing when I add an attribute to them. Why? And are there others that will crash that I haven't found yet?
 * In the customizer, sometimes editing a setting in the advanced section kicks you back to the widget screen immediately; but you might like to make more edits. (See comments below)
 * On this branch, the /wp-admin/widgets.php page loads with the footer collapsed. There is an uncaught `Exception { name: "NS_ERROR_FAILURE", message: "", result: 2147500037,` shown in the browser console that's related to this behavior.
 * Github tests are failing

critterverse commented:
Hi all, core contributor here who was working on widgets recently. This takes up a bit more space visually but I wonder if something like this could work? 

<img width="280" alt="widget-visibility" src="https://user-images.githubusercontent.com/77359364/130527152-961b2c02-0509-4665-b8b5-0a88877dfffe.png">

The inset gray border treatment isn't something we use elsewhere but there's not really a similar control in core. Feel free to adapt or adjust!

jeherve commented:
> Github tests are failing

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,385 @@
+					</div>
+					<div className="widget-vis__select">
+						<SelectControl
+							label="Major Rule"
```
```suggestion
							label={ __( Major Rule', 'jetpack' ) }
```

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,385 @@
+				</div>
+			</div>
+			<div className="widget-vis__rule-col-2">
+				<p className="widget-vis__is">{ __( 'is', 'jetpack' ) }</p>
```
I think we'll need context on this string, like on the PHP version:
https://github.com/Automattic/jetpack/blob/3f8e9de7051fe6b26bcbb861de409a7e5aae1cd1/projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php#L494

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,385 @@
+					<div className="widget-vis__select">
+						<SelectControl
+							className="widget-vis__select-multi-level"
+							label="Minor Rule"
```
```suggestion
							label={ __( 'Minor Rule', 'jetpack' ) }
```

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,385 @@
+const RuleSep = props => {
+	const { isAnd } = props;
+	if ( isAnd ) {
+		return <div className="widget-vis__and-or">{ __( 'and', 'jetpack' ) }</div>;
+	}
+	return <div className="widget-vis__and-or">{ __( 'or', 'jetpack' ) }</div>;
```
Just like above, I think we need to add context to those strings.

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,385 @@
+import { Fragment, useEffect, useCallback, useMemo } from '@wordpress/element';
+import { BaseControl, Button, SelectControl, ToggleControl } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+import { InspectorAdvancedControls } from '@wordpress/block-editor'; // eslint-disable-line import/no-unresolved
```
I must admit that it took me a few seconds to understand where the settings were in the UI. I expected a new panel, as I've learnt to ignore the "Advanced" panel since I've hardly ever had the use for it since I started using the block editor.

That's just my personal experience though :)

mreishus replied:
I was going off Matias' comment in the original issue, however I'd be happy to make it a new panel if that's something that's easily done, do we have any examples of doing that elsewhere?

jeherve replied:
We rely on `InspectorControls` for quite a few of our blocks, but I don't believe we have any example of InspectorControls added to existing core blocks at the moment.

I'm happy to defer to @mtias on this though; I'm not Gutenberg UI expert :)

mreishus replied:
Correct, this feature must be turned on in jetpack settings:
![2021-08-30_10-36](https://user-images.githubusercontent.com/937354/131365212-900d8a03-9839-44f5-bd59-f477d6baad30.png)

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php:
```diff
@@ -63,6 +64,7 @@ public static function init() {
 			// On a screen that is hosting the API in the gutenberg editing experience.
 			if ( is_customize_preview() || 'widgets.php' === $pagenow ) {
 				$add_data_assets_to_page = true;
+				$add_block_controls      = true;
```
Would it make sense to add the settings to the regular block editor too in the future? I feel it could be useful there, although I can see how confusing it would be to have a feature called widget visibility impact and be useful when editing a post or a page.

mreishus replied:
I agree this is a really powerful feature that could be opened up across all contexts of the site; if we go that way I'd like to handle it in a second step with more planning and discussion.

mreishus replied:
We are currently not, and I'm not sure how to add this.

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php:
```diff
@@ -699,25 +728,119 @@ static function generate_condition_key( $rule ) {
+	 * @param array $blocks (By reference; will be modified).
+	 * @return array [$blocks, $indexes_removed]
+	 */
+	public static function recursively_filter_blocks( &$blocks ) {
```
Do you think it would be possible to add unit tests to these new methods?

mreishus replied:
Added unit tests for this method here: 476ed773c17337cbfef55f721ef76e70e6dc4674
Notes:
* They are in the "wrong" directory. `./projects/plugins/jetpack/tests/php/modules/widgets/test_widget-visibility.php`. I wanted to place them in a new directory, ``./projects/plugins/jetpack/tests/php/modules/widget-visibility/`, but I couldn't get them to run inside that directory, even though `widget-visibility` is listed in the `projects/plugins/jetpack/tests/php/test-get-modules.php` file.
* In order to get the tests working, I had to: `jetpack docker sh`, then `cd /tmp/wordpress-develop/tests/phpunit/data`, then `svn up -r44722`, then `cd ../includes/` , then again `svn up -r44722`
  * After this `jetpack docker phpunit` would start running where it previously would not, but it would crash before completing all tests, and leave me without a report of failing tests. However, `jetpack docker phpunit -- --filter=widget` is sufficient to run my new test file without crashing.

jeherve replied:
> then cd /tmp/wordpress-develop/tests/phpunit/data, then svn up -r44722, then cd ../includes/ , then again svn up -r44722

jeherve replied:
> I wanted to place them in a new directory, ./projects/plugins/jetpack/tests/php/modules/widget-visibility/, but I couldn't get them to run inside that directory, even though widget-visibility is listed in the projects/plugins/jetpack/tests/php/test-get-modules.php file.

Try adding the new dir here:

https://github.com/Automattic/jetpack/blob/88ecfe04895c5014ec78f6558fa3030130314466/projects/plugins/jetpack/phpunit.xml.dist#L54-L56

That should do the trick!

jeherve replied:
> Added unit tests for this method here: [476ed77](https://github.com/Automattic/jetpack/commit/476ed773c17337cbfef55f721ef76e70e6dc4674)

Thank you! ❤️

sixhours commented:
> Hi all, core contributor here who was working on widgets recently. This takes up a bit more space visually but I wonder if something like this could work?

@critterverse Heck yeah, thank you so much! Will run with it.

sixhours commented:
Updated the styling and markup:

<img width="279" alt="Screen Shot 2021-08-24 at 11 35 50 AM" src="https://user-images.githubusercontent.com/2124984/130647389-d9ec925e-e0ee-495b-966b-d640fa0bf1a2.png">

This is much cleaner, thanks again @critterverse !

mreishus commented:
> This isn't working on WordPress.com at all now - it's not showing the widget visibility settings under advanced in the customizer for non-legacy widgets. Is this due to the change to use jsx?
Can you share the steps you're using to test on WP.com?

mreishus commented:
Current issues:

 * In the customizer, sometimes editing a setting in the advanced section kicks you back to the widget screen immediately; but you might like to make more edits.
 * Unit tests in wrong directory.
 * Possibly moving visibility settings from advanced panel to its own panel.
 * Testing is not working on WPCOM after adding the es6 build.
 * I cannot add attributes to these types of blocks: 'core/archives', 'core/latest-comments', 'core/latest-posts'. Most blocks let me add a custom attribute, but with these, they start crashing when I add an attribute to them. Why? And are there others that will crash that I haven't found yet?
   * Possibly not a blocker.

sixhours commented:
Something else I've noticed in testing is that the Customizer preview *always* shows the block regardless of what rules I set. I can't recall if this is expected based on the Classic Widgets version of the plugin; I'll test it on a site with Classic Widgets enabled.

mreishus commented:
Oddly I'm not able to reproduce the reversed filtering behavior. Here's me trying with both tag and monthly archive:


https://user-images.githubusercontent.com/937354/130819645-5818f100-7718-4353-b2de-a61843fe138a.mp4


https://user-images.githubusercontent.com/937354/130819654-71dc625c-437a-4e25-ad30-716661804dbf.mp4

sixhours commented:
We figured it out on Slack, I had an old copy of `widget-conditions.php` that was messing me up: p1629906555054900/1629906151.053200-slack-C029HQKQ146

Tested all the major rules against a bunch of minor ones with show/hide variations and got all expected results. 😌

mreishus commented:
> Something else I've noticed in testing is that the Customizer preview _always_ shows the block regardless of what rules I set. I can't recall if this is expected based on the Classic Widgets version of the plugin; I'll test it on a site with Classic Widgets enabled.
> 
> Edit: Confirmed this same thing happens with the classic version of the plugin so I think it's safe to ignore for now.

I think this is a side effect of the plugin disabling the filtering code on any page where you can edit widgets. It doesn't want the widgets to be hidden, causing them to no longer be editable, but it does not distinguish between the left hand portion of the customizer and the preview area.

sixhours commented:
Minor bug: Child pages are supposed to appear indented in the Pages rule, but we show `&nbsp;` instead:

<img width="317" alt="Screen Shot 2021-08-25 at 11 58 14 AM" src="https://user-images.githubusercontent.com/2124984/130825225-77869a64-18bb-41f3-b140-6a4eb7c55240.png">

mtias commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,372 @@
+		if ( rules.length === 0 ) {
+			mainRender = (
+				<BaseControl
+					id="widget-vis__wrapper"
```
Mmm, there seems to be something amiss in how this is structured because the button should not need a label. Probably the right pattern to follow here is https://developer.wordpress.org/block-editor/reference-guides/components/base-control/#basecontrol-visuallabel

mreishus commented on projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php:
```diff
@@ -1044,3 +1198,10 @@ static function migrate_post_type_rules() {
+// Add the 'conditions' attribute to server side rendered blocks
+// 'init' happens too late to hook on block registration.
+global $pagenow;
+if ( is_customize_preview() || 'widgets.php' === $pagenow || ( false !== strpos( $_SERVER['REQUEST_URI'], '/wp-json/wp/v2/block-renderer' ) ) ) {
```
Fixed here: fe9250d09

mreishus commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -0,0 +1,372 @@
+				<BaseControl
+					id="widget-vis__wrapper"
+					className="widget-vis__wrapper"
+					label={ __( 'Visibility Settings', 'jetpack' ) }
```
Fixed here: c05f56dfe37914dd

mreishus commented:
Here are the outstanding issues:

* **(1) Customizer Bug.** F5 -> Visit customizer -> Widgets -> Select a block widget -> Three dots menu -> Show more settings -> Advanced -> Add new rule. 
  * Expect to see: A new rule added and you are still looking at the list of rules.
  * Actually see: A new rule added and you are back to looking at the block widget. You must: Three dots menu -> Show more settings -> Advanced -> Add new rule.
  * **This bug is non-unique to the code in this branch.** It's an existing bug. For example, try:   F5 -> Visit customizer -> Widgets -> Select a block widget -> Three dots menu -> Show more settings -> Set the "Font Size" select to Large.  Now you're brought back to the block widget in the same manner.
  * **This bug is caused by code in core WP.org.** Commenting out [these 8 lines](https://github.com/WordPress/wordpress-develop/blame/d4abd5ca4a500f7b704c59d7a03ac1c583fde6a3/src/js/_enqueues/wp/customize/controls.js#L7199-L7206) stops it from happening. They were added in [this commit](https://github.com/WordPress/wordpress-develop/commit/d2540cfa069c23adc2f214bb40a0bb51eee445d3).
  * I don't think this bug should be solved in this PR, but elsewhere. Should it be a blocker for this feature?
* **(2) Performance monitoring.** @mtias  asks "Are we monitoring any performance changes this might introduce?"
  * We are not, and I'm currently not sure how to handle this. 
* **(3) Github build.** The github build step `Build / Build all projects (pull_request) ` is failing on step `composer build-production`.
  * When I run this locally, I see two types of errors.
    * Some are missing eslint definitions related to instant-search. I'm also seeing this on the master branch.
      * Ignoring these as they are non-unique.
    * Another is `  error  Parsing error: 'import' and 'export' may appear only with 'sourceType: module` on the `dev/jetpack/projects/plugins/jetpack/_inc/build/widget-visibility/editor/util.min.js` file.
      * Util.js is a secondary file I used for one function, `unescape()`. Rather than wrestle with the build system, I simply inlined the the file here: 5f4f424624
* **(4) Dotcom Support.** We have D65619-code, which does not apply to my sandbox for testing (it complains about the files already existing), and is not getting automatically updated. There is also D65978-code (closed), D66022-code (closed), and p3topS-Vt-p2.
  *  Need an updated patch to verify on dotcom.

We're very close, but today is the code freeze, so I'm guessing this isn't making it in.

sdixon194 commented:
@mreishus I'm fine with pushing this to 10.2, however if y'all are really that close and expect to have something ready early enough in the Beta period, I'd consider cherry picking so long as we have adequate time for testing before the release next Tuesday.

mreishus commented:
@jeherve Since only the dotcom patch and a performance review are pending, what do you think about timing here?
Additionally, I will be unavailable Sept 6th-10th if that changes anything.

jeherve commented:
@mreishus Given the timing, I would suggest the following:
- Let's aim to get this PR merged early in the 10.2 release cycle, so any time starting today. :) That will give everyone time to test and refine it / fix bugs in follow-up PRs based on feedback during the month of September, for an official release on Atomic / Jetpack in the 10.2 release in October.
- Let's merge the build infra to Jetpack first (I opened #20910 for that). Once this is merged and you rebase this PR, this should solve the problems you've been having with Fusion, and would also make reviewing this PR easier since it will be smaller.

What do you think?

mreishus commented:
Sounds good. Rebasing..

mreishus commented:
Rebased. 
* :heavy_check_mark: Feature is working on jetpack docker as expected. 
* :heavy_check_mark: D65619 is now applying cleanly and providing the feature to wpcom.
* :x: D65619 is failing to build because of an unrelated problem: `ThemesTest::test_switch_theme_for_purchased_premium_theme_should_succeed()` is failing. Reported to themes team.
* :x: Minor CSS differences on WPCOM.
![2021-09-01_11-13](https://user-images.githubusercontent.com/937354/131707446-b252851e-7ac8-4f6e-b351-c4ce705c5c93.png)
![2021-09-01_11-15](https://user-images.githubusercontent.com/937354/131707526-39d7bd06-0573-46b5-95d5-f04e784bfd0a.png)
* :x: I'm starting to think the customizer bug is a blocker. This feature is annoying to use on WPCOM.

https://user-images.githubusercontent.com/937354/131707648-6a145730-8929-4e2b-b81e-5dd0f4529160.mp4

mreishus commented:
Created gutenberg PR for the customizer bug: https://github.com/WordPress/gutenberg/pull/34543

sixhours commented:
Great news, https://github.com/WordPress/gutenberg/pull/34543 was merged! 🥳 Slated for 11.5.

mreishus commented:
I revisited this and I think we're in good shape. Gutenberg will be releasing the fix to the customizer bug on Wednesday, 9/15, in 11.5. I am not aware of any unresolved issues. Matias has asked if we are measuring potential performance impacts; we are not, and I'm not sure how to.

mreishus commented:
Moved tests to the new directory and verified the same number of assertions is being run from `jetpack docker phpunit -- --filter=widget` (and that that number goes down when the file is emptied).

jeherve commented on projects/plugins/jetpack/modules/widget-visibility/editor/index.jsx:
```diff
@@ -1,9 +1,394 @@
+	'visibilityAdvancedControls'
+);
+
+wp.hooks.addFilter( 'editor.BlockEdit', 'widget/visibility', visibilityAdvancedControls );
```
Here and above, you could have added an import for `addFilter` at the top of the file, it may be easier to quickly go through?

That's not a blocker though.

jeherve reviewed (dismissed):
This tests well for me. I have a couple of nitpicks, but nothing that should block us from merging.

I think it would make sense to have a call for testing once this is enabled on WordPress.com, so other automatticians can give us some feedback. What do you think?

****************

For some select elements, we end up having a bit of padding once the element is selected; this is not really a big issue, but it may be cleaner to remove that extra padding?
<img width="304" alt="Screenshot 2021-09-17 at 15 22 20" src="https://user-images.githubusercontent.com/426388/133789825-7b8b91b6-a581-4839-b0f3-0bf06b5a1e10.png">
![image](https://user-images.githubusercontent.com/426388/133789864-c91fef0c-dd76-49a4-8684-a556de315922.png)

mreishus commented:
Will be happy to merge and put out a call for testing soon, once G 11.5.0 (released yesterday, 2021-09-16) is on wpcom. 

Looking at small bug I just discovered on WPCOM. The "Add new Rule" button is in the wrong place, due to CSS rules loading in a different order across the two environments.

![2021-09-17_10-10_1](https://user-images.githubusercontent.com/937354/133806509-ec51d1ff-e960-4e34-82af-f04320a565e4.png)
![2021-09-17_10-11](https://user-images.githubusercontent.com/937354/133806531-6a69cf9a-40b0-42a0-a367-54a1c37d7aad.png)
![2021-09-17_10-11_1](https://user-images.githubusercontent.com/937354/133806543-0f9196e2-c43f-429b-81e8-ef397d950b5b.png)

mreishus commented:
My current plan is to wait until Gutenberg 11.5.0 is deployed on wpcom (should only be a few days), then merge this, deploy to wpcom, and post a call for testing. Let me know if I should be approaching this a different way.

mreishus commented:
Thanks, @chad1008 . I ran into one extra issue while testing on WPCOM. If I create widgets with visibility rules in the customizer, then go to /wp-admin/widgets.php, those widgets are subject to filtering and might not display on that page. If I make changes and save, then I have just wiped out those hidden widgets.
This is not the case on a standalone install done through jetpack docker, it's happening on WPCOM only. There is already some code to make exceptions for widget.php but it appears not to work correctly on WPCOM. Investigating further..

mreishus commented:
The bypass check is here: 

https://github.com/Automattic/jetpack/blob/3a9cc7a569bc02710351b7d531683c04ac0fc98d/projects/plugins/jetpack/modules/widget-visibility/widget-conditions.php#L762-L766

On my jetpack docker instance, loading widgets.php makes a json call to "sidebars", with these values that pass the above check:
```
    [raw_ref] => http://localhost/wp-admin/widgets.php
    [request_uri] => /wp-json/wp/v2/sidebars?per_page=100&_locale=user
```

On my sandbox, loading widgets.php makes json calls to both "sidebars" and "widgets", with these values that fail the above check:
```
    [raw_ref] => https://public-api.wordpress.com/wp-admin/rest-proxy/                                                                                      
    [request_uri] => /wp/v2/sites/123456/sidebars?_envelope=1&per_page=100&_gutenberg_nonce=deadbeef&_locale=user    
```

```
    [raw_ref] => https://public-api.wordpress.com/wp-admin/rest-proxy/
    [request_uri] => /wp/v2/sites/123456/widgets?_envelope=1&context=edit&per_page=100&_embed=about&_gutenberg_nonce=deadbeef&_locale=user
```

I am currently unsure why my wpcom side requests both "sidebars" and "widgets", while my jetpack docker site requests only "sidebars". I am using the 2021 theme in both.
Also currently unsure how to keep the meaning of "Allow all JSON requests to bypass widget filtering, but only if initiated from widgets.php", since I don't seem to have access to that information in the WPCOM environment. I wonder if allowing all JSON requests to bypass widget filtering would be acceptable, or if it would bring unwanted behavior?

mreishus commented:
Rebased and added a specific wpcom check: https://github.com/Automattic/jetpack/pull/20731/commits/19d9796a18de16d3d23c82740ee3152e6acb51fc.

Here is a video of the problem it is solving:

https://user-images.githubusercontent.com/937354/134254081-0d49d825-77e3-42ff-85be-cdb495f48d7f.mp4

* BEFORE: Visiting widgets.php on wpcom simple would apply filtering rules, meaning that some widgets with visibility rules would not display on this editing page. But while editing, all widgets should display.
* AFTER: Visiting widgets.php on wpcom simple will not apply filtering rules.

Looking for another :heavy_check_mark: and planning to merge/deploy tomorrow if all goes well (since WPCOM is now updated with the latest version of gutenberg). Also, an extra quick verification of 19d9796a18de from either @cpapazoglou or @dsas would be nice :pray: (referring to earlier conversation [here](https://github.com/Automattic/jetpack/pull/20255#discussion_r669469783))

fullofcaffeine commented:
👋🏻 Hi folks, @simison told us you wanted a heads-up when 11.5.x was available on dotcom. It's been [shipped](https://github.com/Automattic/wp-calypso/issues/56098) today to simple and WoA.

dsas reviewed (changes requested):
The most recent changes make sense to me and I can't reproduce the problem described after applying D65619-code to my wpcom sandbox.

I can reproduce a problem with adding widgets via widget.php. If I delete a widget, that is persisted but if I add one or edit a widget, it isn't persisted.
I have  D65619-code applied to my wpcom sandbox and am using my sandbox.
1. go to sitename.wordpress.com/wp-admin/widgets.php
2. Add a widget, any widget
3. Press update
4. Reload the page

What I expected to see was the newly added widget there, what I actually saw was it not there. If I open up the customizer it's not there either.

Here [is a video](https://user-images.githubusercontent.com/93301/134324024-cc74cb20-7bf1-4620-bb86-a955ec4a6aca.mp4), including inspection of some of the API request/response. I'm not sure why this is happening or why it only happens when I have this patch on my sandbox.



I did find a whole bunch of other new widget issues too, but they all happen without this patch applied to my wpcom sandbox, noting them below for awareness but I'll file proper issues for them.
 * [Legacy widgets aren't rendering a preview](https://github.com/Automattic/wp-calypso/issues/56460)
 * The snackbar that displays the 'widgets saved' message on widgets.php is half hidden behind the left hand nav
 * I've managed to get myself API rate limited (429 responses) by just refreshing the widgets.php page a few times

mreishus commented:
I am able to recreate the above issue, but I also see it when my sandbox is completely clean, that is, when no patches are applied to it. So I don't think it's caused by this PR.
`svn revert -R ~/public_html && svn cleanup ~/public_html --remove-unversioned && svn up ~/public_html`
* Visit /wp-admin/widgets.php
* See the current state of my site (1 block)
* Add new paragraph block and press update
* Observe that 2 blocks are POSTed, but in the response to the post, the rendered value of the new block is `''`
* F5 and only see the original block (expect to see: the new block added as well)

dsas reviewed (approved):
Sorry @mreishus you're right, I just checked again and I too can reproduce it on a clean sandbox

mreishus commented:
Rebased and reported the above bug in https://github.com/Automattic/wp-calypso/issues/56509 . I think we're ready to proceed here.

jeherve reviewed (approved):
This looks good to me, let's merge it! 🚢

mreishus commented:
rWP232223

mreishus commented:
This has been reverted due to causing 500 errors. Two separate issues:

* Out of memory: https://github.com/Automattic/wp-calypso/issues/56526
* null->render(): https://github.com/Automattic/wp-calypso/issues/56546
